File: java/src/main/java/com/openadr/client/OADR3.java
package com.openadr.client;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.openadr.client.model.*;
import com.openadr.client.validation.ValidationUtils;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * OpenADR 3 client for Java
 * 
 * This client provides access to all OpenADR 3.1.0 API operations with
 * OAuth2 Client Credentials Flow authentication and comprehensive validation.
 */
public class OADR3 {
    private static final Logger logger = LoggerFactory.getLogger(OADR3.class);
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final OADR3Config config;
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private volatile String accessToken;
    private volatile long tokenExpiresAt;

    /**
     * Creates a new OpenADR 3 client instance
     * 
     * @param config The client configuration
     */
    public OADR3(OADR3Config config) {
        this.config = config;
        
        // Configure HTTP client with reasonable timeouts
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .build();

        // Configure JSON object mapper
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        this.objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
    }

    /**
     * Gets a valid OAuth2 access token, refreshing if necessary
     */
    private synchronized String getAccessToken() throws IOException {
        // Check if we have a valid token (with 30-second buffer)
        long now = System.currentTimeMillis();
        if (accessToken != null && tokenExpiresAt > now + 30000) {
            return accessToken;
        }

        // Request new token using OAuth2 Client Credentials flow
        RequestBody tokenRequestBody = new FormBody.Builder()
                .add("grant_type", "client_credentials")
                .add("client_id", config.getClientId())
                .add("client_secret", config.getClientSecret())
                .add("scope", config.getScope() != null ? config.getScope() : "")
                .build();

        Request tokenRequest = new Request.Builder()
                .url(config.getBaseUrl() + "/auth/token")
                .post(tokenRequestBody)
                .addHeader("Accept", "application/json")
                .build();

        try (Response response = httpClient.newCall(tokenRequest).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("OAuth2 token request failed: " + response.code() + " " + response.message());
            }

            ResponseBody body = response.body();
            if (body == null) {
                throw new IOException("Empty response body from token endpoint");
            }

            String responseString = body.string();
            TokenResponse tokenResponse = objectMapper.readValue(responseString, TokenResponse.class);
            
            this.accessToken = tokenResponse.getAccessToken();
            // Set expiration with 30-second buffer
            this.tokenExpiresAt = now + (tokenResponse.getExpiresIn() - 30) * 1000L;
            
            logger.debug("Successfully obtained OAuth2 access token");
            return accessToken;
        }
    }

    /**
     * Makes an HTTP request with proper authentication and error handling
     */
    private <T> APIResponse<T> makeRequest(String method, String path, Object requestBody, 
                                          TypeReference<T> responseType) throws IOException {
        String token = getAccessToken();
        
        // Build URL
        String url = config.getBaseUrl() + path;
        
        // Build request
        Request.Builder requestBuilder = new Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer " + token)
                .addHeader("Accept", "application/json");

        // Add request body if present
        if (requestBody != null) {
            String jsonBody = objectMapper.writeValueAsString(requestBody);
            RequestBody body = RequestBody.create(jsonBody, JSON);
            
            switch (method.toUpperCase()) {
                case "POST":
                    requestBuilder.post(body);
                    break;
                case "PUT":
                    requestBuilder.put(body);
                    break;
                case "PATCH":
                    requestBuilder.patch(body);
                    break;
                default:
                    throw new IllegalArgumentException("Unsupported method with body: " + method);
            }
        } else {
            switch (method.toUpperCase()) {
                case "GET":
                    requestBuilder.get();
                    break;
                case "DELETE":
                    requestBuilder.delete();
                    break;
                default:
                    throw new IllegalArgumentException("Unsupported method: " + method);
            }
        }

        Request request = requestBuilder.build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            ResponseBody responseBody = response.body();
            String responseString = responseBody != null ? responseBody.string() : "";

            if (response.isSuccessful()) {
                // Parse successful response
                if (responseString.isEmpty() || responseType == null) {
                    return new APIResponse<>(response.code(), null);
                }
                
                T responseData = objectMapper.readValue(responseString, responseType);
                return new APIResponse<>(response.code(), responseData);
            } else {
                // Parse error response
                APIError error;
                try {
                    error = objectMapper.readValue(responseString, APIError.class);
                } catch (Exception e) {
                    // Fallback error
                    error = new APIError("HTTP_ERROR", response.message(), response.code(), responseString);
                }
                return new APIResponse<>(response.code(), error);
            }
        }
    }

    /**
     * Makes a GET request with query parameters
     */
    private <T> APIResponse<T> makeGetRequest(String path, TypeReference<T> responseType) throws IOException {
        return makeRequest("GET", path, null, responseType);
    }

    // Programs API

    /**
     * Searches all programs
     * 
     * @param targets Optional target filters
     * @param skip Number of records to skip for pagination
     * @param limit Maximum number of records to return (max 50)
     * @return API response containing list of programs
     */
    public APIResponse<List<Program>> searchAllPrograms(List<String> targets, Integer skip, Integer limit) 
            throws IOException {
        
        // Validate parameters
        ValidationUtils.validateSearchParams(skip, limit).throwIfInvalid();
        
        // Build query string
        StringBuilder pathBuilder = new StringBuilder("/programs");
        String queryParams = buildQueryString(
                "targets", targets,
                "skip", skip,
                "limit", limit
        );
        if (!queryParams.isEmpty()) {
            pathBuilder.append("?").append(queryParams);
        }

        APIResponse<List<Program>> response = makeGetRequest(pathBuilder.toString(), 
                new TypeReference<List<Program>>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            for (Program program : response.getResponse()) {
                ValidationUtils.validate(program).throwIfInvalid();
            }
        }
        
        return response;
    }

    /**
     * Creates a new program
     * 
     * @param program The program to create
     * @return API response containing the created program
     */
    public APIResponse<Program> createProgram(Program program) throws IOException {
        if (program == null) {
            throw new IllegalArgumentException("Program cannot be null");
        }
        
        // Validate input
        ValidationUtils.validate(program).throwIfInvalid();
        
        APIResponse<Program> response = makeRequest("POST", "/programs", program, 
                new TypeReference<Program>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            ValidationUtils.validate(response.getResponse()).throwIfInvalid();
        }
        
        return response;
    }

    /**
     * Searches for a specific program by ID
     * 
     * @param programId The program ID
     * @return API response containing the program
     */
    public APIResponse<Program> searchProgramByProgramId(String programId) throws IOException {
        ValidationUtils.validateId(programId, "programId").throwIfInvalid();
        
        APIResponse<Program> response = makeGetRequest("/programs/" + programId, 
                new TypeReference<Program>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            ValidationUtils.validate(response.getResponse()).throwIfInvalid();
        }
        
        return response;
    }

    /**
     * Updates an existing program
     * 
     * @param programId The program ID
     * @param program The updated program data
     * @return API response containing the updated program
     */
    public APIResponse<Program> updateProgram(String programId, Program program) throws IOException {
        ValidationUtils.validateId(programId, "programId").throwIfInvalid();
        if (program == null) {
            throw new IllegalArgumentException("Program cannot be null");
        }
        
        // Validate input
        ValidationUtils.validate(program).throwIfInvalid();
        
        APIResponse<Program> response = makeRequest("PUT", "/programs/" + programId, program, 
                new TypeReference<Program>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            ValidationUtils.validate(response.getResponse()).throwIfInvalid();
        }
        
        return response;
    }

    /**
     * Deletes a program
     * 
     * @param programId The program ID
     * @return API response
     */
    public APIResponse<Void> deleteProgram(String programId) throws IOException {
        ValidationUtils.validateId(programId, "programId").throwIfInvalid();
        
        return makeRequest("DELETE", "/programs/" + programId, null, null);
    }

    // Events API

    /**
     * Searches all events
     * 
     * @param programId Optional program ID filter
     * @param skip Number of records to skip for pagination
     * @param limit Maximum number of records to return (max 50)
     * @return API response containing list of events
     */
    public APIResponse<List<Event>> searchAllEvents(String programId, Integer skip, Integer limit) 
            throws IOException {
        
        // Validate parameters
        ValidationUtils.validateSearchParams(skip, limit).throwIfInvalid();
        
        // Build query string
        StringBuilder pathBuilder = new StringBuilder("/events");
        String queryParams = buildQueryString(
                "programId", programId,
                "skip", skip,
                "limit", limit
        );
        if (!queryParams.isEmpty()) {
            pathBuilder.append("?").append(queryParams);
        }

        APIResponse<List<Event>> response = makeGetRequest(pathBuilder.toString(), 
                new TypeReference<List<Event>>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            for (Event event : response.getResponse()) {
                ValidationUtils.validate(event).throwIfInvalid();
            }
        }
        
        return response;
    }

    /**
     * Creates a new event
     * 
     * @param event The event to create
     * @return API response containing the created event
     */
    public APIResponse<Event> createEvent(Event event) throws IOException {
        if (event == null) {
            throw new IllegalArgumentException("Event cannot be null");
        }
        
        // Validate input
        ValidationUtils.validate(event).throwIfInvalid();
        
        APIResponse<Event> response = makeRequest("POST", "/events", event, 
                new TypeReference<Event>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            ValidationUtils.validate(response.getResponse()).throwIfInvalid();
        }
        
        return response;
    }

    // Reports API

    /**
     * Searches all reports
     * 
     * @param programId Optional program ID filter
     * @param clientName Optional client name filter
     * @param skip Number of records to skip for pagination
     * @param limit Maximum number of records to return (max 50)
     * @return API response containing list of reports
     */
    public APIResponse<List<Report>> searchAllReports(String programId, String clientName, 
                                                     Integer skip, Integer limit) throws IOException {
        
        // Validate parameters
        ValidationUtils.validateSearchParams(skip, limit).throwIfInvalid();
        
        // Build query string
        StringBuilder pathBuilder = new StringBuilder("/reports");
        String queryParams = buildQueryString(
                "programId", programId,
                "clientName", clientName,
                "skip", skip,
                "limit", limit
        );
        if (!queryParams.isEmpty()) {
            pathBuilder.append("?").append(queryParams);
        }

        APIResponse<List<Report>> response = makeGetRequest(pathBuilder.toString(), 
                new TypeReference<List<Report>>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            for (Report report : response.getResponse()) {
                ValidationUtils.validate(report).throwIfInvalid();
            }
        }
        
        return response;
    }

    /**
     * Creates a new report
     * 
     * @param report The report to create
     * @return API response containing the created report
     */
    public APIResponse<Report> createReport(Report report) throws IOException {
        if (report == null) {
            throw new IllegalArgumentException("Report cannot be null");
        }
        
        // Validate input
        ValidationUtils.validate(report).throwIfInvalid();
        
        APIResponse<Report> response = makeRequest("POST", "/reports", report, 
                new TypeReference<Report>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            ValidationUtils.validate(response.getResponse()).throwIfInvalid();
        }
        
        return response;
    }

    // VENs API

    /**
     * Searches VENs
     * 
     * @param venName Optional VEN name filter
     * @param skip Number of records to skip for pagination
     * @param limit Maximum number of records to return (max 50)
     * @return API response containing list of VENs
     */
    public APIResponse<List<Ven>> searchVens(String venName, Integer skip, Integer limit) 
            throws IOException {
        
        // Validate parameters
        ValidationUtils.validateSearchParams(skip, limit).throwIfInvalid();
        
        // Build query string
        StringBuilder pathBuilder = new StringBuilder("/vens");
        String queryParams = buildQueryString(
                "venName", venName,
                "skip", skip,
                "limit", limit
        );
        if (!queryParams.isEmpty()) {
            pathBuilder.append("?").append(queryParams);
        }

        APIResponse<List<Ven>> response = makeGetRequest(pathBuilder.toString(), 
                new TypeReference<List<Ven>>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            for (Ven ven : response.getResponse()) {
                ValidationUtils.validate(ven).throwIfInvalid();
            }
        }
        
        return response;
    }

    /**
     * Creates a new VEN
     * 
     * @param ven The VEN to create
     * @return API response containing the created VEN
     */
    public APIResponse<Ven> createVen(Ven ven) throws IOException {
        if (ven == null) {
            throw new IllegalArgumentException("VEN cannot be null");
        }
        
        // Validate input
        ValidationUtils.validate(ven).throwIfInvalid();
        
        APIResponse<Ven> response = makeRequest("POST", "/vens", ven, 
                new TypeReference<Ven>() {});
        
        // Validate response data
        if (response.isSuccess() && response.getResponse() != null) {
            ValidationUtils.validate(response.getResponse()).throwIfInvalid();
        }
        
        return response;
    }

    // Utility methods

    /**
     * Builds a query string from key-value pairs
     */
    private String buildQueryString(Object... params) {
        StringBuilder query = new StringBuilder();
        
        for (int i = 0; i < params.length; i += 2) {
            String key = (String) params[i];
            Object value = params[i + 1];
            
            if (value != null) {
                if (query.length() > 0) {
                    query.append("&");
                }
                
                if (value instanceof List) {
                    List<?> list = (List<?>) value;
                    for (int j = 0; j < list.size(); j++) {
                        if (j > 0) query.append("&");
                        query.append(key).append("=").append(list.get(j));
                    }
                } else {
                    query.append(key).append("=").append(value);
                }
            }
        }
        
        return query.toString();
    }

    /**
     * OAuth2 token response model
     */
    private static class TokenResponse {
        @com.fasterxml.jackson.annotation.JsonProperty("access_token")
        private String accessToken;
        
        @com.fasterxml.jackson.annotation.JsonProperty("token_type")
        private String tokenType;
        
        @com.fasterxml.jackson.annotation.JsonProperty("expires_in")
        private int expiresIn;
        
        @com.fasterxml.jackson.annotation.JsonProperty("scope")
        private String scope;

        public String getAccessToken() {
            return accessToken;
        }

        public void setAccessToken(String accessToken) {
            this.accessToken = accessToken;
        }

        public String getTokenType() {
            return tokenType;
        }

        public void setTokenType(String tokenType) {
            this.tokenType = tokenType;
        }

        public int getExpiresIn() {
            return expiresIn;
        }

        public void setExpiresIn(int expiresIn) {
            this.expiresIn = expiresIn;
        }

        public String getScope() {
            return scope;
        }

        public void setScope(String scope) {
            this.scope = scope;
        }
    }
}

File: node/src/oadr3.ts
import * as Types from './openadr-types.js';
import type { OADR3Config, OAuth2Token, APIResponse } from './types.js';

export class OADR3 {
  private config: OADR3Config;
  private token: OAuth2Token | null = null;

  constructor(config: OADR3Config) {
    this.config = config;
    
    // Ensure baseUrl doesn't end with slash
    if (this.config.baseUrl.endsWith('/')) {
      this.config.baseUrl = this.config.baseUrl.slice(0, -1);
    }
  }

  private async getAccessToken(): Promise<string> {
    // Check if we have a valid token
    if (this.token && this.token.expires_at > Date.now()) {
      return this.token.access_token;
    }

    // Request new token using OAuth2 Client Credentials flow
    const tokenUrl = `${this.config.baseUrl}/auth/token`;
    const body = new URLSearchParams({
      grant_type: 'client_credentials',
      client_id: this.config.clientId,
      client_secret: this.config.clientSecret,
      ...(this.config.scope && { scope: this.config.scope })
    });

    const response = await fetch(tokenUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Accept': 'application/json'
      },
      body: body.toString()
    });

    if (!response.ok) {
      throw new Error(`OAuth2 token request failed: ${response.status} ${response.statusText}`);
    }

    const tokenData = await response.json() as OAuth2Token;
    
    // Calculate expiration time (subtract 30 seconds for buffer)
    tokenData.expires_at = Date.now() + (tokenData.expires_in - 30) * 1000;
    
    this.token = tokenData;
    return tokenData.access_token;
  }

  private async makeRequest<T>(
    method: string,
    path: string,
    params?: Record<string, any>,
    body?: any
  ): Promise<APIResponse<T>> {
    try {
      const accessToken = await this.getAccessToken();
      
      // Build URL with query parameters
      let url = `${this.config.baseUrl}${path}`;
      if (params && Object.keys(params).length > 0) {
        const searchParams = new URLSearchParams();
        for (const [key, value] of Object.entries(params)) {
          if (value !== undefined && value !== null) {
            if (Array.isArray(value)) {
              value.forEach(v => searchParams.append(key, String(v)));
            } else {
              searchParams.append(key, String(value));
            }
          }
        }
        url += `?${searchParams.toString()}`;
      }

      const headers: Record<string, string> = {
        'Authorization': `Bearer ${accessToken}`,
        'Accept': 'application/json'
      };

      if (body && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
        headers['Content-Type'] = 'application/json';
      }

      const response = await fetch(url, {
        method,
        headers,
        ...(body && { body: JSON.stringify(body) })
      });

      let responseData: any;
      const contentType = response.headers.get('content-type');
      
      if (contentType && contentType.includes('application/json')) {
        responseData = await response.json();
      } else {
        responseData = await response.text();
      }

      if (response.ok) {
        return {
          status: response.status,
          response: responseData as T
        };
      } else {
        return {
          status: response.status,
          problem: responseData
        };
      }
    } catch (error) {
      throw new Error(`API request failed: ${error}`);
    }
  }

  // Programs API
  async searchAllPrograms(
    targets?: Types.Target[],
    skip?: number,
    limit?: number
  ): Promise<APIResponse<Types.Program[]>> {
    // Validate inputs
    if (targets) {
      for (const target of targets) {
        const validation = Types.Validators.validateTarget(target);
        if (!validation.valid) {
          throw new Error(`Invalid target: ${JSON.stringify(validation.errors)}`);
        }
      }
    }

    if (skip !== undefined && (skip < 0 || !Number.isInteger(skip))) {
      throw new Error('Skip must be a non-negative integer');
    }

    if (limit !== undefined && (limit < 0 || limit > 50 || !Number.isInteger(limit))) {
      throw new Error('Limit must be an integer between 0 and 50');
    }

    const params: Record<string, any> = {};
    if (targets) params.targets = targets;
    if (skip !== undefined) params.skip = skip;
    if (limit !== undefined) params.limit = limit;

    const result = await this.makeRequest<Types.Program[]>('GET', '/programs', params);
    
    // Validate response if successful
    if (result.response && Array.isArray(result.response)) {
      for (const program of result.response) {
        const validation = Types.Validators.validateProgram(program);
        if (!validation.valid) {
          throw new Error(`Invalid program in response: ${JSON.stringify(validation.errors)}`);
        }
      }
    }

    return result;
  }

  async createProgram(program: Types.ProgramRequest): Promise<APIResponse<Types.Program>> {
    // Validate input
    const validation = Types.Validators.validateProgramRequest(program);
    if (!validation.valid) {
      throw new Error(`Invalid program request: ${JSON.stringify(validation.errors)}`);
    }

    const result = await this.makeRequest<Types.Program>('POST', '/programs', undefined, program);
    
    // Validate response if successful
    if (result.response) {
      const responseValidation = Types.Validators.validateProgram(result.response);
      if (!responseValidation.valid) {
        throw new Error(`Invalid program in response: ${JSON.stringify(responseValidation.errors)}`);
      }
    }

    return result;
  }

  async searchProgramByProgramId(programID: string): Promise<APIResponse<Types.Program>> {
    if (!programID) {
      throw new Error('programID is required');
    }

    const result = await this.makeRequest<Types.Program>('GET', `/programs/${programID}`);
    
    // Validate response if successful
    if (result.response) {
      const validation = Types.Validators.validateProgram(result.response);
      if (!validation.valid) {
        throw new Error(`Invalid program in response: ${JSON.stringify(validation.errors)}`);
      }
    }

    return result;
  }

  async updateProgram(programID: string, program: Types.ProgramRequest): Promise<APIResponse<Types.Program>> {
    if (!programID) {
      throw new Error('programID is required');
    }

    // Validate input
    const validation = Types.Validators.validateProgramRequest(program);
    if (!validation.valid) {
      throw new Error(`Invalid program request: ${JSON.stringify(validation.errors)}`);
    }

    const result = await this.makeRequest<Types.Program>('PUT', `/programs/${programID}`, undefined, program);
    
    // Validate response if successful
    if (result.response) {
      const responseValidation = Types.Validators.validateProgram(result.response);
      if (!responseValidation.valid) {
        throw new Error(`Invalid program in response: ${JSON.stringify(responseValidation.errors)}`);
      }
    }

    return result;
  }

  async deleteProgram(programID: string): Promise<APIResponse<void>> {
    if (!programID) {
      throw new Error('programID is required');
    }

    return await this.makeRequest<void>('DELETE', `/programs/${programID}`);
  }

  // Reports API
  async searchAllReports(
    programID?: string,
    clientName?: string,
    reportRequestID?: string,
    reportSpecifierID?: string,
    skip?: number,
    limit?: number
  ): Promise<APIResponse<Types.Report[]>> {
    // Validate inputs
    if (skip !== undefined && (skip < 0 || !Number.isInteger(skip))) {
      throw new Error('Skip must be a non-negative integer');
    }

    if (limit !== undefined && (limit < 0 || limit > 50 || !Number.isInteger(limit))) {
      throw new Error('Limit must be an integer between 0 and 50');
    }

    const params: Record<string, any> = {};
    if (programID) params.programID = programID;
    if (clientName) params.clientName = clientName;
    if (reportRequestID) params.reportRequestID = reportRequestID;
    if (reportSpecifierID) params.reportSpecifierID = reportSpecifierID;
    if (skip !== undefined) params.skip = skip;
    if (limit !== undefined) params.limit = limit;

    const result = await this.makeRequest<Types.Report[]>('GET', '/reports', params);
    
    // Validate response if successful
    if (result.response && Array.isArray(result.response)) {
      for (const report of result.response) {
        const validation = Types.Validators.validateReport(report);
        if (!validation.valid) {
          throw new Error(`Invalid report in response: ${JSON.stringify(validation.errors)}`);
        }
      }
    }

    return result;
  }

  async createReport(report: Types.ReportRequest): Promise<APIResponse<Types.Report>> {
    // Validate input
    const validation = Types.Validators.validateReportRequest(report);
    if (!validation.valid) {
      throw new Error(`Invalid report request: ${JSON.stringify(validation.errors)}`);
    }

    const result = await this.makeRequest<Types.Report>('POST', '/reports', undefined, report);
    
    // Validate response if successful
    if (result.response) {
      const responseValidation = Types.Validators.validateReport(result.response);
      if (!responseValidation.valid) {
        throw new Error(`Invalid report in response: ${JSON.stringify(responseValidation.errors)}`);
      }
    }

    return result;
  }

  // Events API  
  async searchAllEvents(
    programID?: string,
    targetType?: string,
    targetValues?: string[],
    skip?: number,
    limit?: number
  ): Promise<APIResponse<Types.Event[]>> {
    // Validate inputs
    if (skip !== undefined && (skip < 0 || !Number.isInteger(skip))) {
      throw new Error('Skip must be a non-negative integer');
    }

    if (limit !== undefined && (limit < 0 || limit > 50 || !Number.isInteger(limit))) {
      throw new Error('Limit must be an integer between 0 and 50');
    }

    const params: Record<string, any> = {};
    if (programID) params.programID = programID;
    if (targetType) params.targetType = targetType;
    if (targetValues) params.targetValues = targetValues;
    if (skip !== undefined) params.skip = skip;
    if (limit !== undefined) params.limit = limit;

    const result = await this.makeRequest<Types.Event[]>('GET', '/events', params);
    
    // Validate response if successful
    if (result.response && Array.isArray(result.response)) {
      for (const event of result.response) {
        const validation = Types.Validators.validateEvent(event);
        if (!validation.valid) {
          throw new Error(`Invalid event in response: ${JSON.stringify(validation.errors)}`);
        }
      }
    }

    return result;
  }

  async createEvent(event: Types.EventRequest): Promise<APIResponse<Types.Event>> {
    // Validate input
    const validation = Types.Validators.validateEventRequest(event);
    if (!validation.valid) {
      throw new Error(`Invalid event request: ${JSON.stringify(validation.errors)}`);
    }

    const result = await this.makeRequest<Types.Event>('POST', '/events', undefined, event);
    
    // Validate response if successful
    if (result.response) {
      const responseValidation = Types.Validators.validateEvent(result.response);
      if (!responseValidation.valid) {
        throw new Error(`Invalid event in response: ${JSON.stringify(responseValidation.errors)}`);
      }
    }

    return result;
  }

  // VENs API
  async searchVens(
    venName?: string,
    skip?: number,
    limit?: number
  ): Promise<APIResponse<Types.Ven[]>> {
    // Validate inputs
    if (skip !== undefined && (skip < 0 || !Number.isInteger(skip))) {
      throw new Error('Skip must be a non-negative integer');
    }

    if (limit !== undefined && (limit < 0 || limit > 50 || !Number.isInteger(limit))) {
      throw new Error('Limit must be an integer between 0 and 50');
    }

    const params: Record<string, any> = {};
    if (venName) params.venName = venName;
    if (skip !== undefined) params.skip = skip;
    if (limit !== undefined) params.limit = limit;

    const result = await this.makeRequest<Types.Ven[]>('GET', '/vens', params);
    
    // Validate response if successful
    if (result.response && Array.isArray(result.response)) {
      for (const ven of result.response) {
        const validation = Types.Validators.validateVen(ven);
        if (!validation.valid) {
          throw new Error(`Invalid ven in response: ${JSON.stringify(validation.errors)}`);
        }
      }
    }

    return result;
  }

  async createVen(ven: Types.VenRequest): Promise<APIResponse<Types.Ven>> {
    // Validate input
    const validation = Types.Validators.validateVenRequest(ven);
    if (!validation.valid) {
      throw new Error(`Invalid ven request: ${JSON.stringify(validation.errors)}`);
    }

    const result = await this.makeRequest<Types.Ven>('POST', '/vens', undefined, ven);
    
    // Validate response if successful
    if (result.response) {
      const responseValidation = Types.Validators.validateVen(result.response);
      if (!responseValidation.valid) {
        throw new Error(`Invalid ven in response: ${JSON.stringify(responseValidation.errors)}`);
      }
    }

    return result;
  }

  // Auth API
  async fetchToken(): Promise<APIResponse<OAuth2Token>> {
    return await this.makeRequest<OAuth2Token>('POST', '/auth/token');
  }
}